source/portable/NetworkInterface/MPS3_AN552/NetworkInterface.c

```diff
@@ -0,0 +1,367 @@
+
+    ulMessageLength = xEthernetInterface0.pxEthernetMACDriver->GetRxFrameSize();
+
+    if( ulMessageLength != 0 )
+    {
+        *pxNetworkBuffer = pxGetNetworkBufferWithDescriptor( ulMessageLength,
+                                                             xDescriptorWaitTime );
+
+        if( *pxNetworkBuffer != NULL )
+        {
+            lReceivedBytes = xEthernetInterface0.pxEthernetMACDriver->ReadFrame(
+                ( ( *pxNetworkBuffer )->pucEthernetBuffer ),
+                ulMessageLength );
+
+            if( lReceivedBytes >= 0 )
+            {
+                ( *pxNetworkBuffer )->xDataLength = lReceivedBytes;
+            }
+            else
+            {
+                FreeRTOS_debug_printf( ( "Error: Ethernet driver ReadFrame returned %d \n", lReceivedBytes ) );
+                lReceivedBytes = 0;
+            }
+        }
+        else
+        {
+            FreeRTOS_debug_printf( ( "pxNetworkBuffer = NULL\n" ) );
+            /* No memory available for NetworkBuffer, drop the frame */
+            xEthernetInterface0.pxEthernetMACDriver->ReadFrame(
+                NULL,
+                0 );
+        }
+    }
+
+    return lReceivedBytes;
+}
+/*-----------------------------------------------------------*/
+
+static BaseType_t xLAN91C111_NetworkInterfaceInitialise( NetworkInterface_t * pxInterface )
+{
+    BaseType_t xReturn = pdPASS;
+    UBaseType_t eEthernetLinkStateTimeOutCount = 0;
+    ARM_ETH_MAC_CAPABILITIES xEthernetMACCapabilities;
+    ARM_ETH_LINK_STATE eEthernetLinkState = ARM_ETH_LINK_DOWN;
+    ARM_ETH_LINK_INFO xEthernetLinkInfo;
+    uint32_t ulMACControlArg;
+
+    ( void ) pxInterface;
+
+    if( xRxTaskHandle == NULL )
+    {
+        /* Task that reads incoming Ethernet frames and sends it FreeRTOS
+         * TCP/IP stack. */
+        xReturn = xTaskCreate( prvRxTask,
+                               "EMAC ",
+                               nwRX_TASK_STACK_SIZE,
+                               NULL,
+                               nwETHERNET_RX_HANDLER_TASK_PRIORITY,
+                               &xRxTaskHandle );
+        configASSERT( xReturn != pdFALSE );
+    }
+
+    /* Initialise Ethernet Driver */
+    xEthernetInterface0.pxEthernetMACDriver->Initialize( prvEthernetDriverNotifications );
+    xEthernetInterface0.pxEthernetMACDriver->PowerControl( ARM_POWER_FULL );
+    xEthernetInterface0.pxEthernetMACDriver->SetMacAddress( ( ARM_ETH_MAC_ADDR * ) &pxInterface->pxEndPoint->xMACAddress );
+
+    /* Initialise PHY Driver */
+    xEthernetInterface0.pxEthernetPHYDriver->Initialize(
+        xEthernetInterface0.pxEthernetMACDriver->PHY_Read,
+        xEthernetInterface0.pxEthernetMACDriver->PHY_Write );
+    xEthernetInterface0.pxEthernetPHYDriver->PowerControl( ARM_POWER_FULL );
+    xEthernetInterface0.pxEthernetPHYDriver->SetInterface( xEthernetMACCapabilities.media_interface );
+    xEthernetInterface0.pxEthernetPHYDriver->SetMode( ARM_ETH_PHY_AUTO_NEGOTIATE );
```

Comment:
Here above, before calling `xTaskCreate()`, you test if `xRxTaskHandle` has been set. That is good.
Likewise, it should be tested if the Ethernet and the PHY drivers have been initialised already.

The function `NetworkInterfaceInitialise()` will be called repeatedly until it returns a non-zero value.

In other drivers you will see an enum that describes different phases in the initialisation. Once example for this is the Xilinx Ultrascale driver:
~~~
typedef enum xEMAC_STATE
{
    xEMAC_Init,
    xEMAC_SetupPHY,
    xEMAC_WaitPHY,
    xEMAC_Ready,
    xEMAC_Fatal,
} EMACState_t;
~~~

Reply:
The initialisation method has been refactored based on the recommended use of `enum xEMAC_STATE`.